align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+    
+    def set_icl_datasets(self):
+        icl_datasets = {}
+        incontext_data = self.read_icl_dataset_files()
+        
+        # Add each target to icl_datasets
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            icl_datasets[sys_kdma_name] = []
+            kdma_incontext_data = incontext_data[sys_kdma_name]
+            
+            # Add each examples to icl_datasets
+            for example in kdma_incontext_data:
+                
+                # Get example prompt
+                character_info = outlines_prompts_utils.get_relevant_structured_character_info(example['state'].characters)
+                icl_scenario_description = scenario_state_description_with_relevant_char_info(example['state'], character_info)
+                icl_choices_with_outcomes = {}
+                for choice in example['choices']:
+                    # TODO: Include outcome prediction for ICL examples?
+                    icl_choices_with_outcomes[choice] = {'predicted_outcome':None}
+                icl_prompt = comparative_kdma_score_prediction_prompt(icl_scenario_description,
+                                                                    icl_choices_with_outcomes,
+                                                                    sys_kdma_name)
+                # Get example response
+                icl_response = {}
+                for action, choice, kdma_value in zip(example['actions'], example['choices'], example["kdma_values"]):
+                    # Only include choice if there is a ground truth KDMA value available
+                    if kdma_value is None:
+                        continue
+                    # Predicted scores are 0-10, KDMA values are 0-1
+                    scaled_kdma_value = int(kdma_value * 10)
+                    icl_response[choice] = {}
+                    icl_response[choice]['reasoning'] = self.get_chain_of_thought_reasoning(target_kdma, action,
+                                                                                            example['state'], choice,
+                                                                                            scaled_kdma_value)
+                    icl_response[choice]['score'] = scaled_kdma_value
+
+                # Add example
+                icl_datasets[sys_kdma_name].append({
+                    "prompt": icl_prompt,
+                    "response": icl_response
+                    })
+                
+        self.icl_datasets = icl_datasets
+
+    def get_chain_of_thought_reasoning(self, target_kdma, action, state, choice, expected_value):
```

Comment:
It's possible that Chris' ICL method will pre-generate chain of thought reasoning. I'm not sure how that fits in here (or if it should). I don't know if that end up in a separate file (if so we'd need to match somehow), or in the original dataset, which means we'd need to extend the loading function. Maybe we're already going to have a split in "input-output" ICL datasets and "Chris" ICL datasets (I forgot what the library he's using is called)

For approaches where we don't predict KDMA values (like the base outlines ADM, even though that's not very valuable to us anymore), it's possible that we'd have pre-generated COT **per target**, so this could get really complicated

Additionally, I think the pre-generated COT is really supposed to be model-specific, so we'll need to come up with a better solution for managing these files if that happens

Reply:
Yeah we will definitely have to give this more thought when we come to it. For Chris's approach we likely can't  use `_read_icl_dataset_files()`, we will need a new helper function for a different style of file that allows us to get target-specific reasoning